DEPRECATION.md

```diff
@@ -0,0 +1,91 @@
+    * Implementation change only
+
+  * __Care required__
+
+    * File/class/function names
+      * Including names passed to `Registrable.register`!
+    * Function signatures
+    * Jsonnet config
+    * Model internals that affect shape of saved weights
+
+2. If so, first consider how critical the change is.
+
+  * __Non-critical__
+
+    * Minor name changes
+    * "Cleaner" APIs with no functional benefit
+
+  * __Medium__
+
+    * Name/API that is actively confusing multiple users.
+      * This means Github issues, messages on user channels, etc.
+    * Useful new feature
+
+  * __High__
+
+    * Major bugs
+
+3. Non-critical changes we should happily avoid. Otherwise, let's try to make the change in a backward compatible manner.
+
+  * Options include:
+    * When deprecating, say, a class, leave a shim to its replacement.
+    * A function signature can likely be changed with a keyword argument.
+    * You can decorate with `Registrable.register` multiple times.
+    * Features can be hidden behind flags.
+    * Files can simply be forked. (for extreme cases)
+
+  * Likely no deprecation needs to occur in this case. However, in the spirit of TOOWTDI, one may still mark the old component as deprecated while _not_ removing it. Usages in AllenNLP should be removed so that our warnings are not spammy. Though silencing warnings is an option, we should "eat our own dog food" or we can hardly expect our users to migrate.
+
+4. No backward compatible change can be made.
+
+  * In this situation it's adivsable to consult with other developers. Hopefully they can propose an alternative solution or help mitigate the churn during the change.
+
+## Mechanics
+
+1. Mark the old component as deprecated by adding a `DeprecationWarning`.
+  * This should include whether the component will be removed.
+  * If it must be removed, describe why in an accompanying comment. Link relevant issues.
+  * [Example](https://github.com/allenai/allennlp/blob/cb9651a4c77c10cbd2d76f79b85c6453386dc229/allennlp/modules/text_field_embedders/basic_text_field_embedder.py#L141)
+  * Provide the version and date when we will remove the feature if applicable.
+    * We should support whichever is longest.
+    * The code should live for at least one full minor version and 3 months before removal.
+      * e.g., if you're committing the deprecation to master while version 0.8.4 is out, then it should live throughout version 0.9 and can first be removed in version 0.10.0.
+      * If this isn't possible, consult with other developers. You should have a quite compelling rationale.
+
+2. Remove any AllenNLP usages of the deprecated feature to avoid warnings.
+  * Suppressing warnings should be done rarely. [See here for instance.](https://github.com/allenai/allennlp/blob/9719b5c71207e642276fb1209ea1a4c8467e0792/allennlp/modules/token_embedders/embedding.py#L14)
+
+3. Wait
+
+4. Coordinate your removal PR such that it will go into at least a minor release, i.e. m.n.0.
```

Comment:
I'm not sure it's realistic for people who deprecated a feature to follow up in 3 months--particularly if they are an outside contributor.  What has been happening is that people add an issue to a future milestone (out two full release versions) but we don't have a great process for actually doing the deprecation when the time comes.  There's often other pressures to get the release done promptly

Do we expect the deprecator or the releaser to do the final removal?  I'm not sure.

Reply:
Added a note on using issues and that "another developer" may be the one to handle the removal PR.